kivy/uix/accordion.py

```diff
@@ -264,10 +264,10 @@ def __init__(self, **kwargs):
         fbind('title_args', trigger_title)
         trigger_title()
 
-    def add_widget(self, widget):
+    def add_widget(self, widget, index=0, canvas=None):
```

Comment:
This should be `def add_widget(self, widget, *largs, **kwargs):` and `return self.container.add_widget(widget, *largs, **kwargs)`. Rather than explicitly listing all the args. 

The same for everywhere else in the PR.

Reply:
What about `clear_widgets` and `remove_widget` method? 
What do you think about cases like `actionbar.py`, `bubble.py` and `splitter.py`?

Solution you're suggesting seems like a good half-way solution, but doesn't make headers consistent for all `Widget` subclasses and user still has to unpack arguments from `*args`. 
I'll update pr with your suggestion if you and other developers think that this is a better way to go.

Reply:
>Solution you're suggesting seems like a good half-way solution, but doesn't make headers consistent for all Widget subclasses and user still has to unpack arguments from *args.

I don't understand what you mean by this. Can you give more details or an example as to what you think the problem is here? The reason for not explicitly listing the arguments, is that as soon as something changes in the argument, like we gain a new argument, all these methods will break again. That's why simply catching unneeded positional and keyword argument and forwarding them to the base class is much preferable.

`remove_widget` takes no arguments, so it doesn't need `*largs, **kwargs`. But it also doesn't break anything and is future proof so it's better to not remove them if it already has it. Unless something is broken, it's best to not change it. `clear_widget` does take parameters, so in cases that overwrite it, it needs to be able to accept these additional arguments, because otherwise it is broken. And in those cases `*largs, **kwargs` is preferable than explicitly listing the arguments for the same reasons. Unless you need the actual argument of course.

Reply:
Just the case of unpacking the arguments from `*largs`. When `index` and `canvas` are in `largs` then you do: 
```
index = largs[0]
canvas = largs[1]
```
So this adds two lines in method or one line 
```index, canvas = largs[0], largs[1]```.
and makes arguments position dependent. Also, it makes user to go up to base classes or to read the documentation on kivy.org to find which arguments are in the list when overriding method in more down-the-tree classes (like Splitter). 

1. What do you suggest for `add_widget` case in `splitter.py`? There `index` is declared in `add_widget` method. 
2. In `bubble.py` we have `*l` for `add_widget`. Should we go with `(widget, *largs, **kwargs)`? 
3. What about `actionbar.py` where `item` is used instead of `widget` and `screen.py` where `screen` is used?

> `remove_widget` takes no arguments, so it doesn't need `*largs, **kwargs`
Method `remove_widget` currently takes one argument `widget` and in some cases `*l` is used or `widget, *largs` or `widget, *largs, **kwargs`. 

Method `clear_widget` is defined is some cases with `clear_widgets(self, **kwargs)` or without any arguments, but its argument is `Widget` class is `children=None`. 

For all `remove_widget` occurrences we can do: 
`def remove_widget(self, widget, *largs, **kwargs)`
and for `clear_widgets` 
`def clear_widgets(self, children=None, *largs, **kwargs)`? 

Another thing, through repository `*args` is used more instead of `*largs`, by `*args` is used in 54 cases more then `*largs`. Should we use `*largs` or `*args`? I would go with `*args`.

Reply:
in case you are concerned about having to explicitly unpack arguments, this example shows that it's not required.

```py
class Something:
    def some_method(self, a, b, c=1, d=2):
        print(a, b, c, d)


class SomethingSpecific(Something):
    def some_method(self, a, *args, **kwargs):
        print("woo: {}".format(a))
        super().some_method(a, *args, **kwargs)


Something().some_method(1, 2, 3, 4)

SomethingSpecific().some_method(1, 2, 3, 4)
```
```
gabriel@gryphon:/tmp> python3 arguments.py
1 2 3 4
woo: 1
1 2 3 4
```

Reply:
@tshirtman wrote out pretty much exactly what I was thinking. The only additional point is that if the signature is `def func(a, b, c=True, d=False)` and the subclass uses only keys `a` and `c`. Then I'd write the subclass signature as `def func(a, b, c=True, **kwargs): super().func(a, b, c=c, **kwargs)` as opposed to `def func(a, *args, c=True, **kwargs): super().func(a, *args, c=c, **kwargs)`. Because the latter would fail if you call it like `func(1, 2, 3, 4)`.

Reply:
The case I had in mind with unpacking is if user wants to access `canvas` argument (or `index`) to do some checking and pass new values to super call or to use them in its own sub-class, but method signature of some `Widget` subclass is:
`def add_widget(self, widget, *args, **kwargs):` 

Then user have to do:
`canvas = args[1] if len(args) > 1 else None`
or
`index = args[0] if args else 0`
or change method signature to:
`def add_widget(self, widget, index=0, canvas=None, *args, **kwargs):`
All cases require of user to know position of these arguments in `args` and their default values. When overriding methods in sub-class editors like PyCharm will autocomplete signature from super-class, making user go back to add `index` or `canvas` if they are needed by sub-class. 

If in the future `Widget` or `Layout` or some other class adds a new argument, then user have to unpack it from `args`. On current changes we can add `**kwargs` making those three methods into:
- `def add_widget(self, widget, index=0, canvas=None, **kwargs):`
- `def remove_widget(self, widget, **kwargs):`
- `def clear_widgets(self, children=None, **kwargs):`

which enables us to use `kwargs` to pass/get all future arguments, enables user to pass its own arguments for its own subclasses and have same signature for all classes for which user can read in `Widget` class. To avoid collision of Kivy and user keys in `kwargs` user would have to use a prefix, like `user_`, for its keys.

By reading your suggestions, approach without `index` and `canvas` would make methods look like this:
- `def add_widget(self, widget, *args, **kwargs):`
- `def remove_widget(self, widget, *args, **kwargs):`
- `def clear_widgets(self, children=None, *args, **kwargs):`

for all classes. This includes classes with `index` or `canvas` already in signature but those arguments are not used in override method. This classes are: `BoxLayout`, `Carousel`, `FloatLayout`, `Layout`, `ScrollView`, `Splitter`, `TabbedPanel`.
For `Widget` class we would have:
- `def add_widget(self, widget, index=0, canvas=None, *args, **kwargs):`
- `def remove_widget(self, widget, *args, **kwargs):`
- `def clear_widgets(self, children=None, *args, **kwargs):`

For classes which use only `index`, like `ActionOverflow`, `ActionView`, we would have:
- `def add_widget(self, widget, index=0, *args, **kwargs):`
- or `index` gets unpacked in method's body with `index = args[0] if args else 0`

Whichever solution we settle on, errors like described in https://github.com/kivy/kivy/issues/6956, https://github.com/kivy/kivy/pull/5715, https://github.com/kivy/kivy/pull/5748 won't happen in the future.

Reply:
> or change method signature to:
`def add_widget(self, widget, index=0, canvas=None, *args, **kwargs):`
All cases require of user to know position of these arguments in args and their default values. When overriding methods in sub-class editors like PyCharm will autocomplete signature from super-class, making user go back to add index or canvas if they are needed by sub-class.

Right, the user can indeed spell out these arguments in order if they want to access them in their overriding of the method, and yes, they have to check the signatures of parent methods for that. I don't think the fact that an IDE tries to be helpful and doesn't come up with the exact argument list the user wants should drive our decision, if we go this way, then we'll just spell out everything everywhere, so such generated code won't ever have the `*args` and `**kwargs` argument (unless we eat them up in `Widget.add_widget`, but that's bad practice), and then if we ever add a new argument to the signature of `Widget.add_widget` then the user's generated code will simply break (unless they manually added `*args` and `**kwargs` defensively, to the code generated by pycharm).

I don't think arguments should get unpacked in the body, when needed, unpacking them in the signature is the proper approach. When they are currently unpacked in the signature, but not used, i would advise removing them from the signature to be handled by *args/**kwargs, and passed to the parent method if this one is called, if the parent method is not called, *args and **kwargs shouldn't be accepted, as the extra arguments are not going to be used by anything.